packages/foam-core/src/model/tags.ts

```diff
@@ -0,0 +1,159 @@
+import { FoamWorkspace } from './workspace';
+import { URI } from './uri';
+import { IDisposable } from '../index';
+import { Resource } from './note';
+
+export type TagMetadata = { uri: URI };
+
+export class FoamTags implements IDisposable {
+  public readonly tags: { [key: string]: TagMetadata[] } = {};
+
+  constructor(private readonly workspace: FoamWorkspace) {}
+
+  /**
+   * List of disposables to destroy with the tags
+   */
+  private disposables: IDisposable[] = [];
+
+  /**
+   * Computes all tags in the workspace and keep them up-to-date
+   *
+   * @param workspace the target workspace
+   * @param keepMonitoring whether to recompute the links when the workspace changes
+   * @returns the FoamTags
+   */
+  public static fromWorkspace(
+    workspace: FoamWorkspace,
+    keepMonitoring: boolean = false
+  ): FoamTags {
+    let tags = new FoamTags(workspace);
+
+    Object.values(workspace.list()).forEach(resource =>
+      tags.resolveResource(resource)
+    );
+
+    if (keepMonitoring) {
+      tags.disposables.push(
+        workspace.onDidAdd(resource => {
+          tags.resolveResource(resource);
+        }),
+        workspace.onDidUpdate(change => {
+          tags.updateTagsForUpdatedResource(change.old, change.new);
+        }),
+        workspace.onDidDelete(resource => {
+          tags.updateTagsForRemovedResource(resource);
+        })
+      );
+    }
+    return tags;
+  }
+
+  updateTagsForUpdatedResource(oldVersion: Resource, newVersion: Resource) {
+    if (oldVersion.tags.size === 0 && newVersion.tags.size === 0) {
+      return;
+    }
+
+    this.resolveUpdatedTags(newVersion, oldVersion);
+    this.resolveMovedResource(newVersion, oldVersion);
+  }
+
+  updateTagsForRemovedResource(resource: Resource): void {
+    if (resource.tags.size === 0) {
+      return;
+    }
+
+    resource.tags.forEach(tag => {
+      this.tags[tag] = this.tags[tag].filter(
+        t => !URI.isEqual(t.uri, resource.uri)
+      );
+
+      if (this.tags[tag].length === 0) {
+        delete this.tags[tag];
+      }
+    });
+  }
+
+  resolveResource(resource: Resource): void {
+    resource.tags.forEach(tag => {
+      this.tags[tag] = this.tags[tag] ?? [];
+      this.tags[tag].push({ uri: resource.uri });
```

Comment:
what happens if a resource defines a tag more than once?

e.g.

```
this is a #good day. And I like #good things.
```

Reply:
As far as I can see, the tags on the resource are unique. Did a little test run, and then I also end up with a single URI for a tag. So, feels like this is good as it is?

Reply:
oh yes, that's a good point, we already make tags unique in the `tags` field.

FYI in the future tags won't be just `string`, but 
```
Tag {
  name: string
  range: Range
}
```

Which means that it would be possible to have more than one with the same name within a given resource.
But probably we can enforce the uniqueness when we'll need to do the refactoring here at that point.

So I think we are good here!